parutils/src/lib.rs

```diff
@@ -1,4 +1,185 @@
+/// use tokio::sync::Semaphore;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let semaphore = Arc::new(Semaphore::new(2)); // Limit concurrency to 2 tasks.
+///     let futures = (1..=3).map(|n| async move { Ok::<_, ()>(n) });
+///
+///     let results =
+///         tokio_run_max_concurrency_fold_result_with_semaphore(futures.into_iter(), semaphore)
+///             .await;
+///     assert_eq!(results.unwrap(), vec![1, 2, 3]);
+/// }
+/// ```
+pub async fn tokio_run_max_concurrency_fold_result_with_semaphore<Fut, T, E>(
+    futures_it: impl Iterator<Item = Fut>,
+    max_concurrent: Arc<Semaphore>,
+) -> Result<Vec<T>, ParutilsError<E>>
+where
+    Fut: Future<Output = Result<T, E>> + Send + 'static,
+    T: Send + Sync + 'static,
+    E: Send + Sync + 'static,
+{
+    let handle = tokio::runtime::Handle::current();
+    let mut js: JoinSet<Result<(usize, T), ParutilsError<E>>> = JoinSet::new();
+    for (i, fut) in futures_it.enumerate() {
+        let semaphore = max_concurrent.clone();
+        js.spawn_on(
+            async move {
+                let _permit = semaphore.acquire().await?;
+                let res = fut.await.map_err(ParutilsError::Task)?;
+                Ok((i, res))
+            },
+            &handle,
+        );
+    }
+
+    let mut results: Vec<Option<T>> = Vec::with_capacity(js.len());
+    (0..js.len()).for_each(|_| results.push(None));
+    while let Some(result) = js.join_next().await {
+        let (i, res) = result??;
+        debug_assert!(results[i].is_none());
+        results[i] = Some(res);
+    }
+    debug_assert!(js.is_empty());
+    debug_assert!(results.iter().all(|r| r.is_some()));
+
+    Ok(results.into_iter().map(|r| r.unwrap()).collect())
+}
+
+/// Like tokio_run_max_concurrency_fold_result_with_semaphore but callers can pass in the number
+/// of concurrent tasks they wish to allow and the semaphore is created inside this function scope
+pub async fn tokio_run_max_concurrency_fold_result<Fut, T, E>(
+    futures_it: impl Iterator<Item = Fut>,
+    max_concurrent: usize,
+) -> Result<Vec<T>, ParutilsError<E>>
+where
+    Fut: Future<Output = Result<T, E>> + Send + 'static,
+    T: Send + Sync + 'static,
+    E: Send + Sync + 'static,
+{
+    let semaphore = Arc::new(Semaphore::new(max_concurrent));
+    tokio_run_max_concurrency_fold_result_with_semaphore(futures_it, semaphore).await
+}
+
+/// Like tokio_run_max_concurrency_fold_result_with_semaphore but does not require
+/// that the futures in the iterator parameter produce a Result. The futures may produce any type
+/// that will be returned.
+///
+/// tokio_run_max_concurrency_fold_result_with_semaphore returns an error as soon as any task returns
+/// an error.
+///
+/// This function can be used to ensure that all tasks are finished and returns Vec<Result<>>
+/// if the futures passed in still return results.
+pub async fn tokio_run_max_concurrency_with_semaphore<Fut, T>(
```

Comment:
nit: I am not huge on having these variants of the function (i.e. Fut doesn't return a Result) since they are unused. I would rather add them when needed instead.